src/librbd/internal.cc

```diff
@@ -1850,29 +1851,65 @@ int mirror_image_disable_internal(ImageCtx *ictx, bool force,
       }
       assert(m_bl->length() == (size_t)r);
 
-      if (m_bl->is_zero()) {
+      if (!m_sparse_size) {
+	if (m_bl->is_zero()) {
+	  delete m_bl;
+	  m_throttle->end_op(r);
+	  return;
+	}
+
+	Context *ctx = new C_CopyWrite(m_throttle, m_bl);
+	auto comp = io::AioCompletion::create(ctx);
+
+	// coordinate through AIO WQ to ensure lock is acquired if needed
+	m_dest->io_work_queue->aio_write(comp, m_offset, m_bl->length(),
+					 std::move(*m_bl),
+                                         LIBRADOS_OP_FLAG_FADVISE_DONTNEED);
+      } else {
+        m_throttle->end_op(r);
```

Comment:
sorry, I'm not catch up with you. Can you explain more specifically.

I know the purpose is to gather multipule C_CopyWrite  callbacks as a single xxxx callback, and we want to call end_op() to release the throttle in the single callbacks.

But still, C_CopyWrite is need to call in order to release the corresponding bufferlist, in this case, as well as to decrease the count in C_Gather. Until count == 0 in C_Gather, automatically call the C_Gather callback to end_op(), am I right?

If though, current  C_GatherBase in Context.h can do all those thing? or we need to use a new class inherit from C_GatherBase and add something new, such as override  new_sub() to do this.